src/net/alarm.rs

```diff
@@ -0,0 +1,183 @@
+                }
+            }
+            Change::Alarm(true) => {
+                if handle_is_some {
+                    Some(Action::Trip)
+                } else {
+                    None
+                }
+            }
+            Change::Armed => {
+                if !self.armed && handle_is_some {
+                    Some(Action::Cancel)
+                } else if self.armed && !handle_is_some {
+                    Some(Action::Spawn(self.timeout.millis()))
+                } else {
+                    None
+                }
+            }
+        }
+    }
+
+    /// rearm the alarm
+    pub fn rearm(
+        &self,
+        handle_is_some: bool,
+    ) -> Option<fugit::Duration<u64, 1, 10000_u32>> {
+        if !handle_is_some && self.armed {
+            Some(self.timeout.millis())
+        } else {
+            None
+        }
+    }
+}
+
+/// Stabilizer mqtt alarm listener.
+pub struct Alarm<
+    Stack,
+    Clock,
+    const MESSAGE_SIZE: usize,
+    const MAX_TOPIC_LENGTH: usize,
+> where
+    Stack: TcpClientStack,
+    Clock: embedded_time::Clock,
+{
+    mqtt: minimq::Minimq<Stack, Clock, MESSAGE_SIZE, 1>,
+    subscribed: bool,
+    prefix: String<MAX_TOPIC_LENGTH>,
+}
+
+impl<
+        Stack,
+        Clock,
+        const MESSAGE_SIZE: usize,
+        const MAX_TOPIC_LENGTH: usize,
+    > Alarm<Stack, Clock, MESSAGE_SIZE, MAX_TOPIC_LENGTH>
+where
+    Stack: TcpClientStack,
+    Clock: embedded_time::Clock + Clone,
+{
+    pub fn new(
+        stack: Stack,
+        client_id: &str,
+        prefix: &str,
+        broker: IpAddr,
+        clock: Clock,
+    ) -> Result<Self, minimq::Error<Stack::Error>> {
+        let mqtt = minimq::Minimq::new(broker, client_id, stack, clock)?;
+        let mut prefix = String::from(prefix);
+        prefix.push_str("/alarm").unwrap();
+        Ok(Self {
+            mqtt,
+            subscribed: false,
+            prefix,
+        })
+    }
+
+    /// Update the alarm mqtt interface. Returns Ok(false) if the alarm was renewed.
+    /// (aka if true has been published on the alarm topic)
+    /// Returns Ok(true) if the alarm was triggered.
+    /// Returns None if we only just subscribed or if there was an invalid message on the alarm topic.
+    pub fn update(
+        &mut self,
+    ) -> Result<Option<bool>, minimq::Error<Stack::Error>> {
+        if !self.mqtt.client().is_connected() {
+            // reset subscription if client disconnects
+            self.subscribed = false;
+        }
+        if !self.subscribed && self.mqtt.client().is_connected() {
+            self.subscribed = true;
+            log::info!("Alarm subscribed to {:?}", self.prefix);
+            self.mqtt
+                .client()
+                .subscribe(&[(self.prefix.as_str()).into()], &[])
+                .map(|_| None) // return None if we just subscribed
+        } else {
+            self.mqtt
+                .poll(|_client, _topic, message, _properties| match message {
```

Comment:
I think you can https://docs.rs/serde-json-core/latest/serde_json_core/fn.from_slice.html# this right away. No need to do it yourself.